### PR TITLE
Feature/distributed metadata

### LINKED ISSUE #121: Naming distributed nodes
We should be able to name distributed nodes to easier differentiate them. This would also include host function for looking up nodes by name and returning names when getting all nodes.

### PR DESCRIPTION
**Pull request by kosticmarin**
Resolves #121 

This PR adds a new argument when starting nodes called `tag` which can repeat. The `tag` argument is used
to store additional user defined metadata about the node in a form of a `key=value`, e.g. `--tag name=node01`.

Further we add a two new functions for looking up nodes via a query. The query syntax is like [URL query string](https://en.wikipedia.org/wiki/Query_string).
The first function `exec_lookup_nodes` executes the query on the control node and stores the results in the 
control client, returning a query id and query results size.
The second function `copy_lookup_nodes_results` copies the query results to the guest memory based on the query id.

Finally on the control server we add a parser for the query string.

### CODE REVIEW

File: crates/lunatic-distributed/src/control/client.rs [hunk 1]
```diff
@@ -23,19 +24,23 @@ pub struct Client {
 
 pub struct InnerClient {
     next_message_id: AtomicU64,
+    next_query_id: AtomicU64,
     node_addr: SocketAddr,
     node_name: String,
     control_addr: SocketAddr,
     connection: Connection,
     pending_requests: DashMap<u64, Arc<AsyncCell<Response>>>,
+    pending_queries: DashMap<u64, Vec<u64>>,
```
Comment (withtypes) [inline crates/lunatic-distributed/src/control/client.rs:33]: Why "pending", these are resolved queries. Let's simply call it `node_queries`?

File: crates/lunatic-distributed/src/control/client.rs [hunk 2]
```diff
@@ -23,19 +24,23 @@ pub struct Client {
...
     nodes: DashMap<u64, NodeInfo>,
     node_ids: RwLock<Vec<u64>>,
+    node_metadata: HashMap<String, String>,
 }
 
 impl Client {
     pub async fn register(
         node_addr: SocketAddr,
         node_name: String,
+        node_metadata: HashMap<String, String>,
```
Comment (withtypes) [inline crates/lunatic-distributed/src/control/client.rs:43]: Maybe simply call it node "attributes".

File: crates/lunatic-distributed/src/control/parser.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
+use anyhow::{anyhow, Result};
+use dashmap::mapref::multiple::RefMulti;
+
```
Comment (withtypes) [inline crates/lunatic-distributed/src/control/parser.rs:3]: Do we have some description of this language? What are valid attribute key/values? What operators?

File: src/mode/execution.rs [hunk 4]
```diff
@@ -18,6 +18,22 @@ use lunatic_runtime::{DefaultProcessConfig, DefaultProcessState};
 
 use uuid::Uuid;
 
+/// Parse a single key-value pair
+fn parse_key_val<K, V>(s: &str) -> Result<(K, V)>
```
Comment (withtypes) [inline src/mode/execution.rs:22]: This doesn't use the parser?

**Review by withtypes — COMMENTED** [review]
Looks ok to me. It adds two new guest fns, it's backwards compatible. @bkolobara we can merge it for next release if it looks good to you.

**Review by withtypes — APPROVED** [review]

**Comment by bkolobara** [comment]
Could you just fix the issues after merging the other branch into main and I'm going to merge this one too?

**Comment by kosticmarin** [comment]
@bkolobara Merge issues have been resolved.

**Comment by bkolobara** [comment]
Awesome! Thanks!

**Pull request merged**